File: java/BalancedBinaryTree.java
class Solution {

    /**
     * Verifies whether a given binary tree is height-balanced. 
     * A tree is considered height-balanced if the depth of the two subtrees never differs by more than 1. 
     * @param root the starting node.
     * @return whether the tree is height-balanced, or not.
     */
    public boolean isBalanced(TreeNode root){

        if(root == null){
            return true;
        }

        return Math.abs(treeHeight(root.left) - treeHeight(root.right)) < 2 &&
        isBalanced(root.left) && isBalanced(root.right);

    }

    /**
     * Helper function to get the height of the tree.
     * If the node is null, then the height of the tree is -1
     * Otherwise get the maximum height butween the left and right subtrees and add. 
     * @param node the starting node.
     * @return the height of the three.
     */
    public int treeHeight(TreeNode node){
        
        if(node == null){
            return -1;
        }
        return 1 + Math.max(treeHeight(node.left), treeHeight(node.right));
    }

}

File: python/BalancedBinaryTree.py
from typing import Optional

from python.TreeNode import TreeNode

class Solution:

    # Since the balabnce of the tree checks that at any given node, the height difference is not > 1, use height helper method to check.
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        # Base case, null root.
        if root is None:
            return True
        
        left_balance = self.height(root.left)
        right_balance = self.height(root.right)

        if abs(left_balance - right_balance) > 1:
            return False
        else: return self.isBalanced(root.left) and self.isBalanced(root.right)


    # Basic helper method to get the height of the tree. 
    def height(self, node: Optional[TreeNode]) -> int:
        
        if node is None:
            return 0
        
        left_height = self.height(node.left)
        right_height = self.height(node.right)

        return max(left_height, right_height) + 1
     